### ISSUE TITLE
syntaxcomplete#Complete could fail with an error when quote (') is a keyword

### ISSUE THREAD

**Issue by cwfoo** [vim/vim#7463]
**Describe the bug**

When quote (`'`) is a keyword (`set iskeyword=39`), [`syntaxcomplete#Complete`](https://github.com/vim/vim/blob/v8.2.2129/runtime/autoload/syntaxcomplete.vim#L142-L209) could fail with an error:

```
Error detected while processing function syntaxcomplete#Complete:
line   63:
E15: Invalid expression: *'
```

Similar to #4952 (which was closed, but the underlying bug still exists). The OP of that issue says that he/she sent an email to the maintainer more than a year ago. I have also sent an email some time ago, but I did not receive a reply. The bug has not been fixed, so I am opening a new bug report.

**To Reproduce**
1. Open an empty file without syntax highlighting (e.g. `vim myfile`).
2. Run the following commands:
```
" (1) Set ' as a keyword.
:setlocal iskeyword=39

" (2) Add syntax region for Lisp quotes (e.g. 'hello).
:syn region myQuoted matchgroup=Delimiter start=/'[`']*/ end=/[ \t\n()\[\]";]/me=e-1

" (3)
:setlocal omnifunc=syntaxcomplete#Complete
```
3. Type a single quote character (`'`).
4. Run Omnicomplete: <kbd>Ctrl</kbd><kbd>x</kbd><kbd>Ctrl</kbd><kbd>o</kbd>
5. An error message appears:
```
Error detected while processing function syntaxcomplete#Complete:
line   63:
E15: Invalid expression: *'
```

**Expected behavior**
There should not be any error.

**Additional context**

The error is caused by these lines:

https://github.com/vim/vim/blob/v8.2.2129/runtime/autoload/syntaxcomplete.vim#L204-L205

```
let expr = 'v:val '.(g:omni_syntax_ignorecase==1?'=~?':'=~#')." '^".escape(base, '\\/.*$^~[]').".*'"
let compl_list = filter(deepcopy(compl_list), expr)
```

The problem is that `base` could contain single quotes, in which case the call to `filter` will produce an error because `expr` contains a quote that is not properly escaped.

**Proposed solution**

Replace

```
let expr = 'v:val '.(g:omni_syntax_ignorecase==1?'=~?':'=~#')." '^".escape(base, '\\/.*$^~[]').".*'"
```

with

```
let expr = 'v:val '.(g:omni_syntax_ignorecase==1?'=~?':'=~#')." '^".substitute(escape(base, '\\/.*$^~[]'),"'","''",'g').".*'"
```

This escapes all single quotes in `base`.

**Comment by chrisbra** [comment]
the change looks good to me, although I personally do not like so much nesting. I think this patch should achieve the same:

```patch
diff --git a/runtime/autoload/syntaxcomplete.vim b/runtime/autoload/syntaxcomplete.vim
index 98584b407..f1c01968b 100644
--- a/runtime/autoload/syntaxcomplete.vim
+++ b/runtime/autoload/syntaxcomplete.vim
@@ -179,7 +179,7 @@ function! syntaxcomplete#Complete(findstart, base)
     endif

     " let base = s:prepended . a:base
-    let base = s:prepended
+    let base = substitute(s:prepended, "'", "''", 'g')

     let filetype = substitute(&filetype, '\.', '_', 'g')
     let list_idx = index(s:cache_name, filetype, 0, &ignorecase)

```

ping @dfishburn (hopefully this is the right user)

**Comment by dfishburn** [comment]
Sorry, I never saw the email that was sent a year ago.
I will have a look at the change.

**Comment by dfishburn** [comment]
@cwfoo Please provide text that I can complete, which demonstrates the problem, including filetype and your version of Vim.

**Comment by dfishburn** [comment]
@cwfoo I have uploaded [version 14.0](https://www.vim.org/scripts/script.php?script_id=3172) to www.vim.org.
   Please try it and if you use other filetypes using the syntax complete, please test it out.
   Then I will pass it along to Bram for inclusion.

**Comment by cwfoo** [comment]
@dfishburn Thank you for fixing the problem.

I am not aware of any built-in filetype that demonstrates this problem. For Vim 8.0.1489 (February 2018) and older, the Scheme syntax file caused this error ([scheme.vim in Vim 8.0.1489](https://github.com/vim/vim/blob/v8.0.1489/runtime/syntax/scheme.vim)). The Scheme syntax file was updated in Vim 8.0.1490, and the problem disappeared for Scheme files. Third-party plugins such as https://github.com/wlangstroth/vim-racket still trigger this error when using version 13.0 of syntaxcomplete.vim.

Because I know of no built-in syntax files that cause this error, I have listed the steps to reproduce the problem. Refer to the "To Reproduce" section in my original post above.

I am using Vim in Debian and Ubuntu. Versions: 8.0.0197 and 8.0.1453 respectively.

**Comment by cwfoo** [comment]
@dfishburn FYI there's a small typo in the comments. "begining" -> "beginning"

https://github.com/vim/vim/blob/de4f95b04134f6dfd10f60051a5078fd903a297d/runtime/autoload/syntaxcomplete.vim#L551

**Comment by dfishburn** [comment]
@cwfoo Thanks, just spell checked the whole file.

**Comment by cwfoo** [comment]
Fixed in 9faec4e3d439968e21ad74e917aebb289df8f849.
